docs/source/globus_action_providers.rst

```diff
@@ -408,3 +408,371 @@ should be used in the Datacite test service or the production service.
+    }
+
+However, the query parameters can also be specified in the input document.
+This allows you to dynamically construct the query parameters using information generated or retrieved in the context of a larger Globus Flow.
+For example:
+
+..  code-block:: json
+
+    {
+        "method": "GET",
+        "url": "https://domain.example/datasets",
+        "query-parameters": {
+            "geo": "AMER",
+            "date": "2021-01-17"
+        }
+    }
+
+
+URL substitutions
+.................
+
+..  csv-table::
+
+    "Key name", "``url-substitutions``"
+    "Value type", "JSON keys and values"
+    "Required", "False"
+    "Default", "*None*"
+
+Some HTTP endpoints require a customized URL hierarchy.
+For example, the URL may require an ID to be embedded in the path:
+
+..  code-block:: text
+
+    https://domain.example/dataset/37/status/42/info
+
+When this is required you can specify substitution patterns in the ``url``
+and dynamically provide the correct values in the input document.
+
+Substitution patterns in the URL are wrapped in curly braces like "``{id}``".
+You can specify multiple substitution patterns, and patterns can be repeated.
+Each pattern MUST be referenced as a key in the ``url-substitutions`` object.
+
+If patterns are specified in the URL but are not provided in the ``url-substitutions`` object an error will be returned.
+Similarly, if pattern values are provided in ``url-substitutions`` but are not required then this will also be treated as an error.
+
+The input document below will result in the same URL as the example above,
+but the dataset ID and the status ID can now be dynamically specified during Flow execution.
+
+..  code-block:: json
+
+    {
+        "method": "GET",
+        "url": "https://domain.example/dataset/{dataset-id}/status/{status-id}/info",
+        "url-substitutions": {
+            "dataset-id": "37",
+            "status-id": "42"
+        }
+    }
+
+This feature can be combined with dynamic query parameters, too.
+
+..  note::
+
+    To support this feature, bare curly braces cannot appear in a URL.
+    If you need to have bare curly braces in the URL you can escape them using URL substitution:
+
+    ..  code-block:: json
+
+        {
+            "method": "GET",
+            "url": "https://domain.example/{open-brace}success{close-brace}/",
+            "url-substitutions": {
+                "open-brace": "{",
+                "close-brace": "}"
+            }
+        }
+
+    This will result in the following URL with curly braces embedded:
+
+    ..  code-block:: text
+
+        https://domain.example/{success}/
+
+
+Errors: JSON schema validation
+..............................
+
+Input documents will be validated against the HTTP action provider JSON schema.
+You can get an authoritative copy of the JSON schema by running this command:
+
+..  code-block:: text
+
+    globus-automate action introspect --action-url https://actions.globus.org/http
```

Comment:
Same idea here as above. Although this one is tricky since its dealing specifically with debugging errors with JSON validation. It may be valuable to add this into its own section since this is useful information that applies to all APs and their behavior will more or less be the same.

Reply:
Agree here, too. As above, if this needs to be handled in this PR I'll make the changes but this may be useful in the context of a broader doc update.

Reply:
Done.